packages/circuit-ui/components/Dialog/Dialog.tsx

```diff
@@ -0,0 +1,349 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use client';
+
+import {
+  forwardRef,
+  type HTMLAttributes,
+  type ReactNode,
+  type RefObject,
+  useCallback,
+  useEffect,
+  useRef,
+} from 'react';
+
+import type { Locale } from '../../util/i18n.js';
+import { applyMultipleRefs } from '../../util/refs.js';
+import dialogPolyfill from '../../vendor/dialog-polyfill/index.js';
+import { isEscape } from '../../util/key-codes.js';
+import { useScrollLock } from '../../hooks/useScrollLock/useScrollLock.js';
+import { getFirstFocusableElement } from '../Modal/ModalService.js';
+import { CloseButton } from '../CloseButton/index.js';
+import type { ClickEvent } from '../../types/events.js';
+import { clsx } from '../../styles/clsx.js';
+import { useClickOutside } from '../../hooks/useClickOutside/index.js';
+import { useEscapeKey } from '../../hooks/useEscapeKey/index.js';
+import { useLatest } from '../../hooks/useLatest/index.js';
+
+import classes from './dialog.module.css';
+
+export const ANIMATION_DURATION = 300;
+
+type DataAttribute = `data-${string}`;
+
+export interface DialogProps
+  extends Omit<HTMLAttributes<HTMLDialogElement>, 'children'> {
+  /**
+   * Whether the modal dialog is open or not.
+   */
+  open: boolean;
+  /**
+   * Whether the dialog is modal or not.
+   */
+  isModal?: boolean;
+  /**
+   * Callback when the modal dialog is closed.
+   */
+  onClose?: () => void;
+  /**
+   * Fired when the dialog starts to close.
+   * Use this to trigger animations before the dialog closes.
+   */
+  onCloseStart?: () => void;
+  /**
+   * Text label for the close button for screen readers.
+   * Important for accessibility.
+   */
+  closeButtonLabel?: string;
+  /**
+   * One or more [IETF BCP 47](https://en.wikipedia.org/wiki/IETF_language_tag)
+   * locale identifiers such as `'de-DE'` or `['GB', 'en-US']`.
+   * When passing an array, the first supported locale is used.
+   * Defaults to `navigator.language` in supported environments.
+   */
+  locale?: Locale;
+  /**
+   * Prevent users from closing the modal by clicking/tapping the overlay or
+   * pressing the escape key. Default `false`.
+   */
+  preventClose?: boolean;
+  /**
+   * Enables focusing a particular element in the dialog content and override default behavior
+   */
+  initialFocusRef?: RefObject<HTMLElement>;
+  /**
+   * a ReactNode or a function that returns the content of the modal dialog.
+   */
+  children?:
+    | ReactNode
+    | (({ onClose }: { onClose?: DialogProps['onClose'] }) => ReactNode);
+  /**
+   * The duration of the dialog animation in milliseconds.
+   * If you wish to animate the dialog, provide a value of the animation duration to enable the animation to complete before the dialog closes.
+   * default: 0
```

Comment:
I've started using proper [JS Doc comment syntax](https://jsdoc.app/tags-default) for default values.

```suggestion
   * @default 0
```